sklearn/ensemble/forest.py

```diff
@@ -79,6 +79,56 @@ def _generate_sample_indices(random_state, n_samples):
     return sample_indices
 
 
+def _get_class_distribution(y):
+    """Private function used to fit function.
+    Args: outcome array y
+    Returns: tuple of
+        - classes: list of classes
+        - class_counts: list of count of each class
+        - class_indices: list of indices of each class
+    """
+    if len(y.shape) > 1:
+        if y.shape[1] == 1:
+            y = y.flatten()
+        else:
+            raise ValueError("Balanced random forest not implemented "
+                             "for multi-output")
+
+    classes = np.unique(y)
+    class_indices = [np.nonzero(y == cls)[0] for cls in classes]
+    class_counts = [len(i) for i in class_indices]
+
+    return classes, class_counts, class_indices
+
+
+def _generate_balanced_sample_indices(random_state, balance_data):
+    """
+    Private function used to _parallel_build_trees function.
+    Generates samples according to the balanced random forest method [1],
+    adapted for multi-class, i.e. a bootstrap sample from the minority
+    class and a random sample with replacement of the same size from all
+    other classes.
+    References
+    ----------
+    .. [1] Chen, C., Liaw, A., Breiman, L. (2004) "Using Random Forest to
+           Learn Imbalanced Data", Tech. Rep. 666, 2004
+    """
+    classes, class_counts, class_indices = balance_data
+    min_count = np.min(class_counts)
+    n_class = len(classes)
+
+    random_instance = check_random_state(random_state)
+    sample_indices = np.empty(n_class*min_count, dtype=int)
+
+    for i, cls, count, indices in zip(range(n_class), classes, class_counts,
```

Comment:
`cls` is a reserved keyword I think. You can use `clazz`

Reply:
```suggestion
    for idx, (class_name, (class_count, class_indices)) in enumerate(class_dict.items()):
```

Reply:
Actually, we could do that with Pyhon and `itertools.chain.from_iterable`. I don't know what is the most efficient (vs. preallocation with numpy array).